File: Android/LookSee/app/src/main/java/looksee/angelll/com/models/BusinessMediaHistoryService.kt
package looksee.angelll.com.models

import com.google.gson.Gson

sealed class BusinessMediaHistoryServiceError(message: String) : Exception(message) {
    data object NotSignedIn :
        BusinessMediaHistoryServiceError(
            "You must be signed in before viewing media history.",
        )

    data object TokensUnavailable :
        BusinessMediaHistoryServiceError("Cognito tokens were unavailable.")

    class BadStatus(val code: Int, val responseBody: String) :
        BusinessMediaHistoryServiceError("Media history API error $code: $responseBody")

    data object InvalidResponse :
        BusinessMediaHistoryServiceError("The media-history response was invalid.")
}

interface BusinessMediaHistoryDataSource {
    suspend fun fetchHistory(
        landmarkId: String,
        limit: Int = 25,
        nextToken: String? = null,
    ): BusinessMediaHistoryResponse
}

class BusinessMediaHistoryService internal constructor(
    private val tokenProvider: IdTokenProvider,
    private val httpClient: BusinessHttpClient,
    private val gson: Gson,
) : BusinessMediaHistoryDataSource {
    constructor() : this(
        AmplifyCognitoIdTokenProvider(),
        UrlConnectionBusinessHttpClient(),
        Gson(),
    )

    internal constructor(
        tokenProvider: IdTokenProvider,
        httpClient: BusinessHttpClient,
    ) : this(tokenProvider, httpClient, Gson())

    override suspend fun fetchHistory(
        landmarkId: String,
        limit: Int,
        nextToken: String?,
    ): BusinessMediaHistoryResponse {
        val token = try {
            tokenProvider.idToken()
        } catch (_: BusinessAuthenticationError.NotSignedIn) {
            throw BusinessMediaHistoryServiceError.NotSignedIn
        } catch (_: BusinessAuthenticationError.TokensUnavailable) {
            throw BusinessMediaHistoryServiceError.TokensUnavailable
        }
        val parameters = linkedMapOf("limit" to limit.coerceIn(1, 100).toString())
        nextToken?.trim()?.takeIf(String::isNotEmpty)?.let {
            parameters["nextToken"] = it
        }
        val endpoint = "$LOOKSEE_API_BASE_URL/business/landmarks/" +
            "${encodedPathSegment(landmarkId)}/media-history"
        val response = httpClient.execute(
            BusinessHttpRequest(
                method = "GET",
                url = urlWithQuery(endpoint, parameters),
                authorization = "Bearer $token",
            ),
        )
        if (response.statusCode !in 200..299) {
            throw BusinessMediaHistoryServiceError.BadStatus(
                response.statusCode,
                response.bodyText,
            )
        }
        return try {
            gson.fromJson(response.bodyText, BusinessMediaHistoryResponse::class.java)
                ?: throw BusinessMediaHistoryServiceError.InvalidResponse
        } catch (error: BusinessMediaHistoryServiceError) {
            throw error
        } catch (_: Exception) {
            throw BusinessMediaHistoryServiceError.InvalidResponse
        }
    }
}

File: ios/LookSeeProto/Models/BusinessMediaHistoryService.swift
//
//  BusinessMediaHistoryService.swift
//  LookSeeProto
//
//  Fetches landmark-scoped upload history from the authenticated API.
//

import Foundation
import Amplify
import AWSPluginsCore

enum BusinessMediaHistoryServiceError: LocalizedError {
    case notSignedIn
    case tokensUnavailable
    case invalidURL
    case badStatus(Int, String)

    var errorDescription: String? {
        switch self {
        case .notSignedIn:
            return "You must be signed in before viewing media history."
        case .tokensUnavailable:
            return "Cognito tokens were unavailable."
        case .invalidURL:
            return "The media-history URL could not be created."
        case .badStatus(let code, let body):
            return "Media history API error \(code): \(body)"
        }
    }
}

final class BusinessMediaHistoryService {
    private let baseURL = URL(
        string: "https://7gmn5z3uf2.execute-api.us-east-1.amazonaws.com/dev"
    )!

    private func getCognitoIDToken() async throws -> String {
        let session = try await Amplify.Auth.fetchAuthSession()

        guard session.isSignedIn else {
            throw BusinessMediaHistoryServiceError.notSignedIn
        }

        guard let tokenProvider = session as? AuthCognitoTokensProvider else {
            throw BusinessMediaHistoryServiceError.tokensUnavailable
        }

        let tokens = try tokenProvider.getCognitoTokens().get()
        return tokens.idToken
    }

    func fetchHistory(
        landmarkId: String,
        limit: Int = 25,
        nextToken: String? = nil
    ) async throws -> BusinessMediaHistoryResponse {
        let idToken = try await getCognitoIDToken()

        let endpoint = baseURL
            .appendingPathComponent("business")
            .appendingPathComponent("landmarks")
            .appendingPathComponent(landmarkId)
            .appendingPathComponent("media-history")

        guard var components = URLComponents(
            url: endpoint,
            resolvingAgainstBaseURL: false
        ) else {
            throw BusinessMediaHistoryServiceError.invalidURL
        }

        var queryItems = [
            URLQueryItem(
                name: "limit",
                value: String(max(1, min(limit, 100)))
            )
        ]

        if let nextToken,
           !nextToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            queryItems.append(
                URLQueryItem(name: "nextToken", value: nextToken)
            )
        }

        components.queryItems = queryItems

        guard let url = components.url else {
            throw BusinessMediaHistoryServiceError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue(
            "Bearer \(idToken)",
            forHTTPHeaderField: "Authorization"
        )
        request.setValue(
            "application/json",
            forHTTPHeaderField: "Accept"
        )
        request.timeoutInterval = 30

        let (data, response) = try await URLSession.shared.data(for: request)

        let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
        let responseBody = String(data: data, encoding: .utf8) ?? ""

        guard (200...299).contains(statusCode) else {
            throw BusinessMediaHistoryServiceError.badStatus(
                statusCode,
                responseBody
            )
        }

        return try JSONDecoder().decode(
            BusinessMediaHistoryResponse.self,
            from: data
        )
    }
}
